### PR TITLE
Feature/captions editable json preview

### PR DESCRIPTION
**Pull request by 902seanryan**
Brought in a JSON editor component which allows for editing of the json, as well as displaying in a tree structure complete with collapsible objects. 

Also made small update to clear out the text editor when a caption is added. Prevents a little bit of confusion

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,15 +11,16 @@
     "bellhop-iframe": "^2.1.2",
     "lodash-es": "^4.17.10",
     "quill": "^1.3.6",
+    "springroll": "git+https://github.com/SpringRoll/SpringRoll.git#v2",
```
Comment (JonGardner-RedSpace) [inline package.json]: should this still be targeting the github instead of npm repo?

File: src/class/CaptionManager.js [hunk 2]
```diff
@@ -117,6 +134,7 @@ class CaptionManager {
 
   emitData() {
     EventBus.$emit('caption_data', this.data);
+    console.log(this.data);
```
Comment (JonGardner-RedSpace) [inline src/class/CaptionManager.js]: `console log`

File: src/components/TextEditor.vue [hunk 3]
```diff
@@ -157,10 +162,12 @@ export default {
 
   mounted() {
     EventBus.$on('caption_changed', this.onUpdate);
+    //EventBus.$on('caption_content_update', this.updateContent);
```
Comment (JonGardner-RedSpace) [inline src/components/TextEditor.vue]: should this be commented out still?
I see it being unregistered below still.

Comment (902seanryan) [reply src/components/TextEditor.vue]: Yeah that was left in accidentally, removed all of it.

**Comment by 902seanryan** [comment]
Found bug with editor sometimes duplicating the first caption object. Seems to happen when a caption is updated via JSON when it isn't the currently selected caption.

**Review by JonGardner-RedSpace — CHANGES_REQUESTED** [review]
Few minor things.

File: src/components/FileDirectory.vue [hunk 4]
```diff
@@ -44,13 +63,13 @@ export default {
   },
   computed: {
     files() {
-      return this.directory.files.map(file => {
+      return this.directory.files.map((file) => {
```
Comment (chipbell4) [inline src/components/FileDirectory.vue]: This looks like we're still using the old directory-based upload than file based. Maybe we need to get `master` merged in?

**Review by chipbell4 — CHANGES_REQUESTED** [review]
Aside from the small thing I flagged, I'm also noticing behavior where the app doesn't update the text area if the JSON is updated. I'm wondering if there's a way for us:
1) Set the current caption being edited in the text box via what JSON section is being edited
2) Update the text in the text box based on what's currently being input via the JSON editor

**Comment by chipbell4** [comment]
Looks like there's some conflicts here as well

**Comment by chipbell4** [comment]
So, JSON editing is looking pretty good. However, it looks like editing the JSON updates the "Text Preview" field, but still doesn't update the "Text Editor" field.

**Review by amalley — CHANGES_REQUESTED** [review]
When editing the JSON, if one of the fields are cleared then that change isn't reflected in the text editor. This includes the content, start, and end fields. Additionally, when a field is cleared and you switch files and back, the original values appear in the JSON before it was cleared.

I am also getting really strange behaviour when trying to add captions with a start and end time of 0. If I don't select to use the text editor values I get an error about the time needing to be positive. We should default these times to 0 when attempting to add a new caption to avoid this error because the text editor shows them as 0. Secondly, trying to add a caption with start and end times of 0 after selecting the "use current time" doesn't add a caption to the JSON preview. However, the text preview "previous" and "next" buttons behave as if a caption was added.

**Comment by 902seanryan** [comment]
> When editing the JSON, if one of the fields are cleared then that change isn't reflected in the text editor. This includes the content, start, and end fields. Additionally, when a field is cleared and you switch files and back, the original values appear in the JSON before it was cleared.
> 
> I am also getting really strange behaviour when trying to add captions with a start and end time of 0. If I don't select to use the text editor values I get an error about the time needing to be positive. We should default these times to 0 when attempting to add a new caption to avoid this error because the text editor shows them as 0. Secondly, trying to add a caption with start and end times of 0 after selecting the "use current time" doesn't add a caption to the JSON preview. However, the text preview "previous" and "next" buttons behave as if a caption was added.
------

these were largely issues with how the caption studio behavior was implemented originally. When a new caption was added, it was set as 0 start and end times, and with a single white space as the content. Which the JSON preview filtered out since the content was 'empty', and the start was less than or equal to the end. 
I've updated how the JSON preview filters data. 

As a note: @chipbell4 This now means that if there are any errors in the caption, then the CaptionManager will print out some errors to the console along with the errors printed out on the page. I don't think this is a big issue but it is worth noting.

**Review by amalley — APPROVED** [review]

**Review by chipbell4 — APPROVED** [review]

**Review by chipbell4 — APPROVED** [review]

**Pull request merged**